src/math/math_base.h

```diff
@@ -105,5 +106,73 @@ struct PointBase3D : public PointBase2D<_Type>
     _Type z;
 };
 
+template <typename _Type>
+struct LineBase2D
+{
+    LineBase2D(const PointBase2D<_Type> & _p1, const PointBase2D<_Type> & _p2)
+        : p1( _p1 ), p2( _p2 )
+    { }
+
+    LineBase2D(const PointBase2D<_Type> & _p, double angleRadians)
+    {
+        this->p1 = PointBase2D<_Type>(_p.x, _p.y);
+        this->p2 = PointBase2D<_Type>(std::cos(angleRadians) * length(), std::sin(angleRadians) * length());
+    }
+
+    bool operator == ( const LineBase2D & line ) const
+    {
+        return p1 == line.p1 && p2 == line.p2;
+    }
+
+    LineBase2D & operator += ( const PointBase2D<_Type> & point )
+    {
+        return LineBase2D(this->p1 + point, this->p2 + point);
+    }
+
+    double length() const
+    {
+        const double x = static_cast<double>(p2.x) - static_cast<double>(p1.x);
+        const double y = static_cast<double>(p2.y) - static_cast<double>(p1.y);
+        return std::sqrt(x * x + y * y);
+    }
+
+    double angle() const
+    {
+        if (p1 == p2) throw std::invalid_argument("Points p1 and p2 must be different!");
+        const double x = static_cast<double>(p2.x) - static_cast<double>(p1.x);
+        const double y = static_cast<double>(p2.y) - static_cast<double>(p1.y);
+        const double angleRadians = std::atan2(y, x);
+        return angleRadians < 0 ? angleRadians + 2 * pvmath::pi : angleRadians;
+    }
+
+    bool intersect(const LineBase2D & line, PointBase2D<_Type> * point) const
+    {
+        // based on Graphics Gems III, Faster Line Segment Intersection, p. 199-202
+        // http://www.realtimerendering.com/resources/GraphicsGems/gems.html#gemsiii
+        const PointBase2D<_Type> a = p2 - p1;
+        const PointBase2D<_Type> b = line.p1 - line.p2;
+        const PointBase2D<_Type> c = p1 - line.p1;
+
+        const double denominator = a.y * b.x - a.x * b.y;
+        if (pvmath::isEqual<_Type>(denominator, 0))
+        {
+            return false; // parallel
+        }
+
+        const double inverse_denominator = 1.0 / denominator;
+        const double na = (b.y * c.x - b.x * c.y) * inverse_denominator;
+        if (point)
+        {
+            *point = p1 + a * na;
+            return true;
+        }
+
+        return false;
+    }
+
+    PointBase2D<_Type> p1;
+    PointBase2D<_Type> p2;
```

Comment:
Let's assume in this case that this kind of vector :) This is not a bug, just expressing my thoughts